embedded-timestamp-mismatch/rewrite_embedded_bucket_id_control_min_mismatch.js

```diff
@@ -0,0 +1,213 @@
+const bucketsColl = db.getCollection('system.buckets.' + collName);
+
+//
+// NON-MODIFIABLE CODE BELOW
+//
+// ------------------------------------------------------------------------------------
+// The "temp" collection should not exist prior to running the script. This will
+// be used for storing the measurements of the buckets with mismatched embedded
+// bucket id timestamp and control.min timestamp.
+// ------------------------------------------------------------------------------------
+listCollectionsRes = db.runCommand({
+                         listCollections: 1.0,
+                         filter: {name: 'temp'}
+                       }).cursor.firstBatch;
+if (listCollectionsRes.length != 0) {
+  print(
+      'Collection `temp` should not exist prior to running the script. Rename or drop the collection before running this script');
+  exit(1);
+}
+
+// ---------------------------------------------------------------------------------------
+// The script will, for each bucket in the affected time-series collection:
+// 1) Detect if the bucket has a mismatch between the embedded bucket id
+// timestamp and the control min timestamp.
+// 2) Re-insert the measurements of the timestamp-mismatched bucket
+// transactionally.
+//    a) Unpack the measurements
+//    b) Repack the measurements into new buckets.
+//    c) Delete the original, problematic bucket from the collection.
+// 3) Validate that there are no buckets with a mismatch between the embedded
+// bucket id timestamp and the control min timestamp.
+// ----------------------------------------------------------------------------------------
+let bucketColl;
+let tsOptions;
+let tempTimeseriesColl;
+let tempTimeseriesBucketsColl;
+
+function setUp() {
+  bucketColl = db.getCollection('system.buckets.' + collName);
+
+  // Create a temp collection to store measurements from the buckets with
+  // mismatched embedded bucket id timestamp and control.min timestamp.
+  tsOptions =
+      db.runCommand({listCollections: 1.0, filter: {name: coll.getName()}})
+          .cursor.firstBatch[0]
+          .options.timeseries;
+
+  db.createCollection('temp', {timeseries: tsOptions});
+  tempTimeseriesColl = db.getCollection('temp');
+  tempTimeseriesBucketsColl = db.getCollection('system.buckets.temp');
+}
+
+// Main function.
+function runFixEmbeddedBucketIdControlMinMismatchProcedure() {
+  setUp();
+  let cursor = bucketsColl.find({}, {_id: true, control: true});
+
+  // Mismatched timestamp buckets will have different types for their
+  // control.min.parameter and control.max.parameter due to type ordering.
+  // Iterate through all buckets, checking if the control.min and control.max
+  // types match. If they do not match, re-insert the bucket.
+  while (cursor.hasNext()) {
+    const bucket = cursor.next();
+    const oidTimestamp = bucket._id.getTimestamp();
+    const controlMinTimestamp = bucket.control.min.t
+
+    if (oidTimestamp != controlMinTimestamp) {
```

Comment:
This should take extended range dates into account. We shouldn't reinsert all data with extended range.

Reply:
Done, and added testing with using extended range docs from timeseries_filter_extended_range.js and ensuring that they don't get overwritten.